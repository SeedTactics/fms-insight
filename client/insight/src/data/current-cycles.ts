/* Copyright (c) 2022, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { LazySeq } from "@seedtactics/immutable-collections";
import { addSeconds } from "date-fns";
import {
  EstimatedCycleTimes,
  isOutlierAbove,
  PartAndStationOperation,
  splitElapsedTimeAmongChunk,
} from "../cell-status/estimated-cycle-times.js";
import { stat_name_and_num } from "../cell-status/station-cycles.js";
import { ActionType, ICurrentStatus, PalletLocation } from "../network/api.js";
import { durationToMinutes, durationToSeconds } from "../util/parseISODuration.js";

export type CurrentCycle = {
  readonly station: string;
  readonly start: Date;
  readonly expectedEnd: Date;
  readonly isOutlier: boolean;
  readonly parts: ReadonlyArray<{ readonly part: string; readonly oper: string }>;
};

function machiningCurrentCycles(
  currentSt: ICurrentStatus,
  estimated: EstimatedCycleTimes,
  palToLoc: ReadonlyMap<string, PalletLocation>
): LazySeq<CurrentCycle> {
  return LazySeq.of(currentSt.material)
    .collect((m) => {
      if (m.action.type !== ActionType.Machining) return null;
      if (!m.location.pallet) return null;
      const loc = palToLoc.get(m.location.pallet);
      if (!loc) return null;
      return { mat: m, loc };
    })
    .groupBy(
      ({ loc }) => loc.group,
      ({ loc }) => loc.num
    )
    .map(([[statGroup, statNum], mats]) => {
      // all mats currently machining at the same station should all have the same part and program
      const stats = estimated.get(
        new PartAndStationOperation(
          mats[0].mat.partName,
          mats[0].mat.process,
          statGroup,
          mats[0].mat.action.program ?? ""
        )
      );
      const elapsedSec = durationToSeconds(mats[0].mat.action.elapsedMachiningTime ?? "PT0S");
      const remainingSec = durationToSeconds(mats[0].mat.action.expectedRemainingMachiningTime ?? "PT0S");
      return {
        station: stat_name_and_num(statGroup, statNum),
        start: addSeconds(currentSt.timeOfCurrentStatusUTC, -elapsedSec),
        expectedEnd: addSeconds(currentSt.timeOfCurrentStatusUTC, remainingSec),
        isOutlier: stats ? isOutlierAbove(stats, elapsedSec / 60 / mats.length) : false,
        parts: [{ part: mats[0].mat.partName, oper: mats[0].mat.action.program ?? "" }],
      };
    });
}

function loadCurrentCycles(
  currentSt: ICurrentStatus,
  estimated: EstimatedCycleTimes,
  palToLoc: ReadonlyMap<string, PalletLocation>
): LazySeq<CurrentCycle> {
  return LazySeq.of(currentSt.material)
    .collect((m) => {
      if (
        m.action.type === ActionType.UnloadToCompletedMaterial ||
        m.action.type === ActionType.UnloadToInProcess
      ) {
        if (!m.location.pallet) return null;
        const loc = palToLoc.get(m.location.pallet);
        if (!loc) return null;
        return { mat: m, material: [m], proc: m.process, path: m.path, loc };
      } else if (m.action.type === ActionType.Loading) {
        if (!m.action.loadOntoPallet) return null;
        const loc = palToLoc.get(m.action.loadOntoPallet);
        if (!loc) return null;
        return {
          mat: m,
          material: [m],
          proc: m.action.processAfterLoad ?? m.process,
          path: m.action.pathAfterLoad ?? m.path,
          loc,
        };
      }
      return null;
    })
    .map((m) => {
      const job = currentSt.jobs[m.mat.jobUnique];
      const pathData = job?.procsAndPaths?.[m.proc - 1]?.paths?.[m.path - 1];
      if (m.mat.action.type === ActionType.Loading) {
        return { ...m, expectedLoadSecs: durationToSeconds(pathData?.expectedLoadTime ?? "PT0S") };
      } else {
        return { ...m, expectedLoadSecs: durationToSeconds(pathData?.expectedUnloadTime ?? "PT0S") };
      }
    })
    .groupBy(
      ({ loc }) => loc.group,
      ({ loc }) => loc.num
    )
    .map(([[statGroup, statNum], mats]) => {
      const matsWithTime = splitElapsedTimeAmongChunk(
        mats,
        (m) => durationToMinutes(m.mat.action.elapsedLoadUnloadTime ?? "PT0S"),
        (m) => m.expectedLoadSecs / 60
      );
      let outlier = false;
      let expectedSecs = 0;
      for (const { cycle: m, elapsedForSingleMaterialMinutes } of matsWithTime) {
        const stats = estimated.get(
          new PartAndStationOperation(
            m.mat.partName,
            m.proc,
            statGroup,
            m.mat.action.type === ActionType.Loading ? "LOAD" : "UNLOAD"
          )
        );
        if (stats && isOutlierAbove(stats, elapsedForSingleMaterialMinutes)) {
          outlier = true;
        }
        if (m.expectedLoadSecs > 0) {
          expectedSecs += m.expectedLoadSecs;
        } else if (stats?.expectedCycleMinutesForSingleMat) {
          expectedSecs += stats.expectedCycleMinutesForSingleMat * 60;
        }
      }
      const elapsedSec = durationToSeconds(mats[0].mat.action.elapsedLoadUnloadTime ?? "PT0S");
      return {
        station: stat_name_and_num(statGroup, statNum),
        start: addSeconds(currentSt.timeOfCurrentStatusUTC, -elapsedSec),
        expectedEnd: addSeconds(currentSt.timeOfCurrentStatusUTC, -elapsedSec + expectedSecs),
        isOutlier: outlier,
        parts: LazySeq.of(mats)
          .distinctAndSortBy(
            (m) => m.mat.partName,
            (m) => m.proc,
            (m) => (m.mat.action.type === ActionType.Loading ? "LOAD" : "UNLOAD")
          )
          .map((m) => ({
            part: m.mat.partName + "-" + m.proc.toString(),
            oper: m.mat.action.type === ActionType.Loading ? "LOAD" : "UNLOAD",
          }))
          .toRArray(),
      };
    });
}

export function currentCycles(
  currentSt: ICurrentStatus,
  estimated: EstimatedCycleTimes
): ReadonlyArray<CurrentCycle> {
  const palToLoc = LazySeq.ofObject(currentSt.pallets).buildHashMap<string, PalletLocation>(
    ([p]) => p,
    (_old, [, pal]) => pal.currentPalletLocation
  );

  return machiningCurrentCycles(currentSt, estimated, palToLoc)
    .concat(loadCurrentCycles(currentSt, estimated, palToLoc))
    .toRArray();
}
