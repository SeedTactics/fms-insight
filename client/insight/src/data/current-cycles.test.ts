/* Copyright (c) 2022, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { PartAndStationOperation, StatisticalCycleTime } from "../cell-status/estimated-cycle-times.js";
import { it, expect } from "vitest";
import { HashMap } from "@seedtactics/immutable-collections";
import { currentCycles } from "./current-cycles.js";
import {
  ActionType,
  ActiveJob,
  ICurrentStatus,
  InProcessMaterial,
  InProcessMaterialAction,
  InProcessMaterialLocation,
  LocType,
  PalletLocation,
  PalletLocationEnum,
  PalletStatus,
  ProcessInfo,
  ProcPathInfo,
} from "../network/api.js";

function fakePalAtMachine(pal: number, num: number): PalletStatus {
  return new PalletStatus({
    palletNum: pal,
    fixtureOnPallet: "",
    onHold: false,
    currentPalletLocation: new PalletLocation({
      loc: PalletLocationEnum.Machine,
      group: "MC",
      num,
    }),
    numFaces: 1,
  });
}

function fakePalAtLoad(pal: number, num: number): PalletStatus {
  return new PalletStatus({
    palletNum: pal,
    fixtureOnPallet: "",
    onHold: false,
    currentPalletLocation: new PalletLocation({
      loc: PalletLocationEnum.LoadUnload,
      group: "L/U",
      num,
    }),
    numFaces: 1,
  });
}

function fakeMC({
  part,
  proc,
  path,
  pal,
  prog,
  elapsedMin,
  remainMin,
}: {
  part: string;
  proc: number;
  path?: number;
  pal: number;
  prog: string;
  elapsedMin: number;
  remainMin: number;
}): InProcessMaterial {
  return new InProcessMaterial({
    materialID: 100,
    jobUnique: "unused-job-unique",
    partName: part,
    process: proc,
    path: path ?? 1,
    signaledInspections: [],
    action: new InProcessMaterialAction({
      type: ActionType.Machining,
      program: prog,
      elapsedMachiningTime: `PT${elapsedMin}M`,
      expectedRemainingMachiningTime: `PT${remainMin}M`,
    }),
    location: new InProcessMaterialLocation({
      type: LocType.OnPallet,
      palletNum: pal,
    }),
  });
}

function fakeLoad({
  uniq,
  part,
  proc,
  path,
  ty,
  pal,
  elapsedMin,
}: {
  uniq: string;
  part: string;
  proc: number;
  path?: number;
  ty: "load" | "unload" | "completed";
  pal: number;
  elapsedMin: number;
}): InProcessMaterial {
  return new InProcessMaterial({
    materialID: 100,
    jobUnique: uniq,
    partName: part,
    process: ty === "load" ? 123094345 : proc,
    path: ty === "load" ? 1239829345 : path ?? 1,
    signaledInspections: [],
    action: new InProcessMaterialAction({
      type:
        ty === "load"
          ? ActionType.Loading
          : ty === "unload"
          ? ActionType.UnloadToInProcess
          : ActionType.UnloadToCompletedMaterial,
      elapsedLoadUnloadTime: `PT${elapsedMin}M`,
      loadOntoPalletNum: ty === "load" ? pal : undefined,
      processAfterLoad: ty === "load" ? proc : undefined,
      pathAfterLoad: ty === "load" ? path ?? 1 : undefined,
    }),
    location:
      ty === "load"
        ? new InProcessMaterialLocation({
            type: LocType.Free,
          })
        : new InProcessMaterialLocation({
            type: LocType.OnPallet,
            palletNum: pal,
          }),
  });
}

it("calculates current cycles", () => {
  const now = new Date(Date.UTC(2018, 2, 5, 7, 0, 0));

  const curSt: ICurrentStatus = {
    timeOfCurrentStatusUTC: now,
    jobs: {},
    pallets: {
      1: fakePalAtLoad(1, 10),
      2: fakePalAtLoad(2, 20),
      3: fakePalAtMachine(3, 6),
      4: fakePalAtMachine(4, 2),
      5: fakePalAtMachine(5, 1),
      6: fakePalAtLoad(6, 30),
    },
    material: [
      // one cycle just getting started
      fakeMC({ part: "part1", proc: 1, pal: 3, prog: "abcd", elapsedMin: 10, remainMin: 18 }),
      fakeMC({ part: "part1", proc: 1, pal: 3, prog: "abcd", elapsedMin: 10, remainMin: 18 }),
      fakeMC({ part: "part1", proc: 1, pal: 3, prog: "abcd", elapsedMin: 10, remainMin: 18 }),

      // one cycle just passed completed
      fakeMC({ part: "part1", proc: 1, pal: 4, prog: "abcd", elapsedMin: 30, remainMin: -2 }),
      fakeMC({ part: "part1", proc: 1, pal: 4, prog: "abcd", elapsedMin: 30, remainMin: -2 }),
      fakeMC({ part: "part1", proc: 1, pal: 4, prog: "abcd", elapsedMin: 30, remainMin: -2 }),

      // one cycle long passed completed
      fakeMC({ part: "part1", proc: 1, pal: 5, prog: "abcd", elapsedMin: 50, remainMin: -22 }),
      fakeMC({ part: "part1", proc: 1, pal: 5, prog: "abcd", elapsedMin: 50, remainMin: -22 }),
      fakeMC({ part: "part1", proc: 1, pal: 5, prog: "abcd", elapsedMin: 50, remainMin: -22 }),

      // loading but very short
      fakeLoad({ uniq: "uniq3", part: "part3", proc: 1, pal: 1, ty: "load", elapsedMin: 10 }),
      fakeLoad({ uniq: "uniq3", part: "part3", proc: 1, pal: 1, ty: "load", elapsedMin: 10 }),
      fakeLoad({ uniq: "uniq4", part: "part4", proc: 2, pal: 1, ty: "unload", elapsedMin: 10 }),
      fakeLoad({ uniq: "uniq4", part: "part4", proc: 2, pal: 1, ty: "completed", elapsedMin: 10 }),

      // loading but at the expected time
      fakeLoad({ uniq: "uniq3", part: "part3", proc: 1, pal: 2, ty: "load", elapsedMin: 20 }),
      fakeLoad({ uniq: "uniq3", part: "part3", proc: 1, pal: 2, ty: "load", elapsedMin: 20 }),
      fakeLoad({ uniq: "uniq4", part: "part4", proc: 2, pal: 2, ty: "unload", elapsedMin: 20 }),
      fakeLoad({ uniq: "uniq4", part: "part4", proc: 2, pal: 2, ty: "completed", elapsedMin: 20 }),

      // loading far past the expected time
      fakeLoad({ uniq: "uniq3", part: "part3", proc: 1, pal: 6, ty: "load", elapsedMin: 50 }),
      fakeLoad({ uniq: "uniq3", part: "part3", proc: 1, pal: 6, ty: "load", elapsedMin: 50 }),
      fakeLoad({ uniq: "uniq4", part: "part4", proc: 2, pal: 6, ty: "unload", elapsedMin: 50 }),
      fakeLoad({ uniq: "uniq4", part: "part4", proc: 2, pal: 6, ty: "completed", elapsedMin: 50 }),
    ],
    alarms: [],
    queues: {},
  };

  expect(currentCycles(curSt, HashMap.empty())).toMatchSnapshot("current cycles without jobs or expected");

  const expected = HashMap.from<PartAndStationOperation, StatisticalCycleTime>([
    [
      new PartAndStationOperation("part1", "MC", "abcd"),
      {
        medianMinutesForSingleMat: 28 / 3, // 3 parts per face
        MAD_aboveMinutes: 1.2,
        MAD_belowMinutes: 1.2,
        expectedCycleMinutesForSingleMat: 28,
      },
    ],
    [
      new PartAndStationOperation("part3", "L/U", "LOAD-1"),
      {
        medianMinutesForSingleMat: 8 / 2, // 2 parts per face
        MAD_aboveMinutes: 1,
        MAD_belowMinutes: 1,
        expectedCycleMinutesForSingleMat: 8 / 2,
      },
    ],
    [
      new PartAndStationOperation("part4", "L/U", "UNLOAD-2"),
      {
        medianMinutesForSingleMat: 11 / 2, // two parts per face
        MAD_aboveMinutes: 1,
        MAD_belowMinutes: 1,
        expectedCycleMinutesForSingleMat: 11 / 2,
      },
    ],
  ]);

  expect(currentCycles(curSt, expected)).toMatchSnapshot("current cycles without jobs, just expected");

  const jobs = {
    // uniq3 only has loading on proc1
    uniq3: new ActiveJob({
      unique: "uniq3",
      partName: "part3",
      routeStartUTC: new Date(),
      routeEndUTC: new Date(),
      archived: false,
      copiedToSystem: true,
      procsAndPaths: [new ProcessInfo({ paths: [{ expectedLoadTime: "PT4.5M" } as ProcPathInfo] })],
    }),
    // uniq4 is unloading on proc2
    uniq4: new ActiveJob({
      unique: "uniq4",
      partName: "part4",
      routeStartUTC: new Date(),
      routeEndUTC: new Date(),
      archived: false,
      copiedToSystem: true,
      procsAndPaths: [
        {} as ProcessInfo,
        new ProcessInfo({ paths: [{ expectedUnloadTime: "PT6M" } as ProcPathInfo] }),
      ],
    }),
  };

  expect(currentCycles({ ...curSt, jobs }, expected)).toMatchSnapshot(
    "current cycles with jobs and expected",
  );
});
