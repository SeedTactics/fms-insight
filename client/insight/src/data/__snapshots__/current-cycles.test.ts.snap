// Vitest Snapshot v1

exports[`calculates current cycles > current cycles with jobs and expected 1`] = `
[
  {
    "expectedEnd": 2018-03-05T07:18:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:50:00.000Z,
    "station": "MC #6",
  },
  {
    "expectedEnd": 2018-03-05T06:58:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:30:00.000Z,
    "station": "MC #2",
  },
  {
    "expectedEnd": 2018-03-05T06:38:00.000Z,
    "isOutlier": true,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:10:00.000Z,
    "station": "MC #1",
  },
  {
    "expectedEnd": 2018-03-05T07:11:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:50:00.000Z,
    "station": "L/U #10",
  },
  {
    "expectedEnd": 2018-03-05T07:01:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:40:00.000Z,
    "station": "L/U #20",
  },
  {
    "expectedEnd": 2018-03-05T06:31:00.000Z,
    "isOutlier": true,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:10:00.000Z,
    "station": "L/U #30",
  },
]
`;

exports[`calculates current cycles > current cycles without jobs or expected 1`] = `
[
  {
    "expectedEnd": 2018-03-05T07:18:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:50:00.000Z,
    "station": "MC #6",
  },
  {
    "expectedEnd": 2018-03-05T06:58:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:30:00.000Z,
    "station": "MC #2",
  },
  {
    "expectedEnd": 2018-03-05T06:38:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:10:00.000Z,
    "station": "MC #1",
  },
  {
    "expectedEnd": 2018-03-05T06:50:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:50:00.000Z,
    "station": "L/U #10",
  },
  {
    "expectedEnd": 2018-03-05T06:40:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:40:00.000Z,
    "station": "L/U #20",
  },
  {
    "expectedEnd": 2018-03-05T06:10:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:10:00.000Z,
    "station": "L/U #30",
  },
]
`;

exports[`calculates current cycles > current cycles without jobs, just expected 1`] = `
[
  {
    "expectedEnd": 2018-03-05T07:18:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:50:00.000Z,
    "station": "MC #6",
  },
  {
    "expectedEnd": 2018-03-05T06:58:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:30:00.000Z,
    "station": "MC #2",
  },
  {
    "expectedEnd": 2018-03-05T06:38:00.000Z,
    "isOutlier": true,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1",
      },
    ],
    "start": 2018-03-05T06:10:00.000Z,
    "station": "MC #1",
  },
  {
    "expectedEnd": 2018-03-05T07:09:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:50:00.000Z,
    "station": "L/U #10",
  },
  {
    "expectedEnd": 2018-03-05T06:59:00.000Z,
    "isOutlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:40:00.000Z,
    "station": "L/U #20",
  },
  {
    "expectedEnd": 2018-03-05T06:29:00.000Z,
    "isOutlier": true,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "start": 2018-03-05T06:10:00.000Z,
    "station": "L/U #30",
  },
]
`;
